### DISCUSSION TITLE
Reporting multiple detections as one?

### DISCUSSION THREAD

**Discussion by Svardsten53** [mcguirepr89/BirdNET-Pi#100] [Ideas]
Would it be possible to report several detections for a species within the same 15 seconds interval as just one detection? Frequent birds tend to otherwise fill the "Today's Detections" list. See example below.

The disadvantage of only reporting one hit may be that, for example, you would not be aware that several birds would fly in the sky (migratory birds). However, one can never be sure if it is sounds from several individuals or just one that has sounded several times.

This is just an idea from me, others probably have other views.

![Skärmbild 2021-12-16 121728](https://user-images.githubusercontent.com/62435779/146364883-045ffe2d-b1ce-42fd-a0e3-2e9e9b86330e.png)

**Comment by mcguirepr89** [comment]
Hello, my friend! Great question

>Would it be possible to report several detections for a species within the same 15 seconds interval as just one detection? Frequent birds tend to otherwise fill the "Today's Detections" list. See example below.

While I'm sure this would be possible with edits to `analyze.py`, I won't be implementing anything like that for BirdNET-Pi since its 24/7 monitoring is designed to take into account everything that can be detected. And while I understand your desire to have a "Today's Detections" will fewer redundancies, the database itself exists to keep these extensive records.

I will put together a script for you, however, that will cull the database of its duplicate entries. Once you've been able to test the script to be sure it does what you'd like, we can set it up to cull the database of duplicates at regular intervals so that your "Today's Detections" view can more closely resemble what you have in mind.

I will get back to you when the script is ready to test and we'll go from there ~~

My best regards as always,
Patrick

**Reply by Svardsten53** [reply]
Thanks Patrick, that sounds very good.

**Reply by mcguirepr89** [reply]
Hello again, Lars-Åke!

Thinking on this more, the desired effect is going to be a high hurdle. 
I would be very happy to be shown an error in my thinking or a simpler way to do things, but this is how I am viewing your request:
Also, any further mention hereafter of the _table_ is in direct reference to the _`detections`_ table, unless referenced as the _`copy table`_

The MySQL query would need to perform the following in some capacity depending on whether the desire is to _remove_ the table entries, or simply display things more selectively:

1. First, select all table entries where there is only one species detection for that timestamp
2. Place the results of step 1 into a copy table
3. Select all table entries where there is more than one detection for a timestamp
4. From the results of step 3, append One(1) table entry, presumably with the highest value Confidence score per species per timestamp to the copy table from step 2.
5. Return the full copy table sorted by time, descending

I suppose there could be a MySQL query that would bang this all out, but I unfortunately cannot conquer that myself just yet. 

I've found plenty of ways to only display entries for which there is only one detection per species per timestamp, but have not been able to get those to also include _any_ of the dections wherein more than one detection is made per species per timestamp.

I will not close my mind on this idea, but something will need to change for me to be able to make much meaningful progress.

I do apologize that I offered a solution too quickly. If I find a way to achieve this for you, I will be certain to let you know immediately.

Kind regards and contrition,
Patrick McGuire

**Reply by Svardsten53** [reply]
Thank you for answering and for investigating. Of course, we will not implement complicated solutions for a simple little issue. It was just an idea from me. I have worked with most things EXCEPT databases, so did not realize that it would be complex.

**Reply by Molokai** [reply]
Maybe I can help as I'm not unfamiliar with databases.
Here is an initial stab - this query groups all detections of the same bird on the same time stamp, showing the highest confidence as well as the number of detections on that time stamp:

```
SELECT d1.date, d1.time, d1.sci_name, d1.com_name, MAX(confidence), COUNT(*) 
FROM detections d1 
GROUP BY d1.date, d1.time, d1.sci_name, d1.com_name 
ORDER BY COUNT(*) DESC;
```

If you want to more easily visualize what it is doing, you can limit it to the highest number of detections:

```
SELECT d1.date, d1.time, d1.sci_name, d1.com_name, MAX(confidence), COUNT(*) 
FROM detections d1
GROUP BY d1.date, d1.time, d1.sci_name, d1.com_name
ORDER BY COUNT(*) DESC
LIMIT 1;

+------------+----------+-------------+------------+-----------------+----------+
| date       | time     | sci_name    | com_name   | max(confidence) | count(*) |
+------------+----------+-------------+------------+-----------------+----------+
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |        0.995919 |        5 |
+------------+----------+-------------+------------+-----------------+----------+
1 row in set (0.001 sec)
```

And then you can see the individual detections that were "clumped together":

```
MariaDB [birds]> select * from detections where date = '2022-02-02' and time = '16:10:31' and sci_name= 'Parus Major';
+------------+----------+-------------+------------+------------+--------+---------+--------+------+------+---------+
| Date       | Time     | Sci_Name    | Com_Name   | Confidence | Lat    | Lon     | Cutoff | Week | Sens | Overlap |
+------------+----------+-------------+------------+------------+--------+---------+--------+------+------+---------+
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |   0.942908 | 59.835 | 10.6739 |    0.7 |    5 | 1.25 |       0 |
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |   0.995363 | 59.835 | 10.6739 |    0.7 |    5 | 1.25 |       0 |
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |   0.993745 | 59.835 | 10.6739 |    0.7 |    5 | 1.25 |       0 |
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |   0.885245 | 59.835 | 10.6739 |    0.7 |    5 | 1.25 |       0 |
| 2022-02-02 | 16:10:31 | Parus major | Kjøttmeis  |   0.995919 | 59.835 | 10.6739 |    0.7 |    5 | 1.25 |       0 |
+------------+----------+-------------+------------+------------+--------+---------+--------+------+------+---------+
5 rows in set (0.001 sec)
```
And here is the skeleton of a view on detections that you can query against (I omitted some of the columns as they aren't all that relevant for the discussion, but you might want to include them):

```
CREATE VIEW grouped_detections AS 
SELECT d1.Date, d1.Time, d1.Sci_Name, d1.Com_Name, MAX(confidence) as Confidence, COUNT(*) as Count
FROM detections d1
GROUP BY d1.date, d1.time, d1.sci_name, d1.com_name

MariaDB [birds]> SELECT * FROM grouped_detections ORDER BY Count DESC LIMIT 5;
+------------+----------+---------------------+------------+------------+-------+
| Date       | Time     | Sci_Name            | Com_Name   | Confidence | Count |
+------------+----------+---------------------+------------+------------+-------+
| 2022-02-02 | 16:10:31 | Parus major         | Kjøttmeis  |   0.995919 |     5 |
| 2022-01-30 | 10:21:58 | Cyanistes caeruleus | blåmeis    |   0.975571 |     4 |
| 2022-02-03 | 05:24:52 | Strix aluco         | Kattugle   |   0.895485 |     4 |
| 2022-01-30 | 10:33:43 | Sitta europaea      | spettmeis  |   0.966406 |     4 |
| 2022-02-02 | 16:10:46 | Parus major         | Kjøttmeis  |   0.996441 |     3 |
+------------+----------+---------------------+------------+------------+-------+
5 rows in set (0.002 sec)
```

I might be able to help out with other queries if there are particular needs.

Hope this Helps!

**Reply by Molokai** [reply]
Oh and if you want a new table with deduplicated detections, you could create it like so:

```
CREATE TABLE deduplicated_detections
SELECT Date, Time, Sci_Name, Com_Name, MAX(confidence) AS Confidence, Lat, Lon, Cutoff, Week, Sens, Overlap, COUNT(*) AS Count
FROM detections
GROUP BY date, time, sci_name, com_name, Lat, Lon, Cutoff, Week, Sens, Overlap;
```

This adds the column Count that indicates how many detections were "collapsed" into one. This assumes that  Lat, Lon, cutoff, Week and Sens fields are identical for all the entries.

I would like to come up with a query that can be run periodically to add new deduplicated entries. If there was a combined date/time field I could just have done this:

```
INSERT INTO deduplicated_detections
SELECT Date, Time, Sci_Name, Com_Name, MAX(confidence) AS Confidence, Lat, Lon, Cutoff, Week, Sens, Overlap, COUNT(*) AS Count
FROM detections
WHERE DateTime > (SELECT MAX(DateTime) FROM deduplicated_detections)
GROUP BY date, time, sci_name, com_name, Lat, Lon, Cutoff, Week, Sens, Overlap;
```
It is probably possible to combine the date and time fields in order to do something similar. I'll take a closer look when I have more time.

Ah this was really quite easy to solve!
```
INSERT INTO deduplicated_detections
SELECT Date, Time, Sci_Name, Com_Name, MAX(confidence) AS Confidence, Lat, Lon, Cutoff, Week, Sens, Overlap, COUNT(*) AS Count
FROM detections
WHERE TIMESTAMP(date,time) > (SELECT MAX(TIMESTAMP(date,time)) FROM deduplicated_detections)
GROUP BY date, time, sci_name, com_name, Lat, Lon, Cutoff, Week, Sens, Overlap;
```

**Reply by mcguirepr89** [reply]
@Molokai This is awesome!! You've given me a lot to work with. Thank you so so much and I will probably be following up with lots of questions :)

My best regards,
Patrick

**Reply by Molokai** [reply]
Fire away! Happy to help :+1:

**Reply by mcguirepr89** [reply]
I've just [pushed the change](https://github.com/mcguirepr89/BirdNET-Pi/commit/f21a48cd7b3ed824026f4311318d03128246ceb8) to `viewdb.php` that now uses:
```
SELECT Date, Time, Sci_Name, Com_Name, MAX(Confidence) 
FROM detections 
WHERE Date = CURDATE() 
GROUP BY Date, Time, Sci_Name, Com_Name 
ORDER BY Time DESC;
```

to just show the highest confidence and "Report Multiple Detections As One"

Thank you again, @Molokai!!!

@Svardsten53 The community solved this for us!! :) :) :)

**Reply by Svardsten53** [reply]
Super! I have included the new code in my own customization and can see the difference.
